ocaml/xapi/xapi_vdi.ml

```diff
@@ -473,6 +473,15 @@ let update_allowed_operations_internal ~__context ~self ~sr_records ~pbd_records
          )
   in
   let all = Db.VDI.get_record_internal ~__context ~self in
+  let vbd_records =
+    match vbd_records with
+    | None when Pool_role.is_master () ->
+        all.Db_actions.vDI_VBDs
+        |> List.rev_map (fun self -> Db.VBD.get_record_internal ~__context ~self)
+        |> Option.some
+    | v ->
+        v
+  in
   let allowed =
     let check x =
```

Comment:
I'm puzzled as to why isn't `check` implemented with a fold_left and not a simple filter.

Reply:
not only that, but it also builds a result value and list for each element, only to throw it away. No O(N^2) problems there, but yes the whole function could be tweaked a bit.

Reply:
I'm tweaking this to use a Set instead, which only needs a few lines changed. I'll measure how much this'd improve the code.
E.g. using sets we can avoid List.intersect, and can use the more efficient Set intersection.

A lot of lists in xapi should be maps or sets, it better models the underlying problem too (e.g. we know the list won't have duplicates, or that duplicates in the list wouldn't make sense, and that the order in the list doesn't matter.... all of which indicate the Set as a more natural choice for a data structure)

Reply:
I've added a benchmark, the O(n^2) improvement results in ~18x improvement on the benchmark. The other changes are smaller (they reduce the amount of allocations a bit, and maybe ~3%).
The improvements might be larger if I used larger lists, but it takes ~15s just to initialize the benchmark as it is. Need to optimize VM/VBD/VDI construction too...